exporters/otlp/src/OpenTelemetry/Exporter/OTLP/Span.hs

```diff
@@ -248,6 +270,176 @@ putWarningLn :: (MonadIO m) => String -> m ()
+-}
+data ExportTraceError = ExportTraceError
+  { rejectedSpans :: Int64
+  , errorMessage :: Text
+  }
+  deriving (Show)
+
+
+instance Exception ExportTraceError where
+  displayException ExportTraceError {..} =
+    "Error: OpenTelemetry Collector rejected " <> show rejectedSpans <> " data points with message: " <> T.unpack errorMessage
+
+
+{- |
+Internal helper.
+Get the connection configuration from the 'OTLPExporterConfig'.
+-}
+grpcConnParams :: OTLPExporterConfig -> G.ConnParams
+grpcConnParams conf =
+  G.def
+    { G.connCompression = grcpCompression conf
+    , G.connDefaultTimeout = grpcTimeout conf
+    }
+
+
+{- |
+Internal helper.
+Get a negationation strategy for the compression format from the 'OTLPExporterConfig'.
+-}
+grcpCompression :: OTLPExporterConfig -> G.Negotation
+grcpCompression conf =
+  case fromMaybe None (otlpCompression conf <|> otlpTracesCompression conf) of
+    None -> G.none
+    GZip -> G.only G.gzip
+
+
+{- |
+Internal helper.
+Get a timeout from the 'OTLPExporterConfig'.
+-}
+grpcTimeout :: OTLPExporterConfig -> Maybe G.Timeout
+grpcTimeout conf =
+  case otlpTimeout conf <|> otlpTracesTimeout conf of
+    Just timeoutMilli
+      | timeoutMilli >= 1 -> toTimeout timeoutMilli
+      | timeoutMilli == 0 -> Nothing
+    _otherwise -> toTimeout defaultExporterTimeout
+  where
+    toTimeout :: Int -> Maybe G.Timeout
+    toTimeout = Just . G.Timeout G.Millisecond . G.TimeoutValue . fromIntegral
+
+
+{- |
+Internal helper.
+Get a server configuration  from the 'OTLPExporterConfig'.
+-}
+grpcServer :: (MonadIO m) => OTLPExporterConfig -> m G.Server
+grpcServer conf = do
+  grpcAddress conf <&> \case
+    serverAddress
+      | otlpInsecure conf || otlpSpanInsecure conf ->
+          G.ServerInsecure serverAddress
+      | otherwise ->
+          G.ServerSecure (grpcServerValidation conf) G.SslKeyLogNone serverAddress
+
+
+{- |
+Internal helper.
+Get a server validation configuration  from the 'OTLPExporterConfig'.
+-}
+grpcServerValidation :: OTLPExporterConfig -> G.ServerValidation
+grpcServerValidation conf =
+  case otlpCertificate conf <|> otlpTracesCertificate conf of
+    Nothing -> G.NoServerValidation
+    Just certificatePath -> G.ValidateServer (G.certStoreFromPath certificatePath)
+
+
+{- |
+Internal helper.
+Get an address  from the 'OTLPExporterConfig'.
+-}
+grpcAddress :: (MonadIO m) => OTLPExporterConfig -> m G.Address
+grpcAddress conf =
+  case otlpEndpoint conf <|> otlpTracesEndpoint conf of
+    Just endpoint
+      | Just uri <- parseURI endpoint
+      , Just uriAuthority <- uriAuthority uri
+      , Just addressPort <- readMaybe (uriPort uriAuthority) ->
+          pure $
+            G.Address
+              { G.addressHost = "localhost"
+              , G.addressPort = addressPort
+              , G.addressAuthority = Nothing
+              }
+      | otherwise -> do
+          putWarningLn $ "Warning: could not parse endpoint '" <> endpoint <> "'"
```

Comment:
I don't think we should be just randomly polluting stderr like this. I see we do it already though.